### PR TITLE
e2e test for L2 Blob 

### PR DESCRIPTION
**Pull request by ping-ke**
**Description**
This PR address the issue(https://github.com/ethstorage/optimism/issues/102) of L2 Blob's E2E tests
For related manual step, see `Test L2 BLOB` at https://perfect-amphibian-929.notion.site/SWC-Devnet-1282e493e88380648486d4961eb91a8d

**Tests**
```
cd op-e2e
make devnet-allocs
go test -run '^(TestSubmitTXWithBlobsFunctionSuccess)$' github.com/ethereum-optimism/optimism/op-e2e/blobs  -count=1 -v
```

**Questions**

Should we add a test case for `using SGT as the gas fee to send TX with blobs`?

### CODE REVIEW

File: op-e2e/faultproofs/util.go [hunk 1]
```diff
@@ -26,7 +29,7 @@ func WithBatcherStopped() faultDisputeConfigOpts {
 
 func WithBlobBatches() faultDisputeConfigOpts {
 	return func(cfg *e2esys.SystemConfig) {
-		cfg.DataAvailabilityType = batcherFlags.BlobsType
+		cfg.DataAvailabilityType = flags.BlobsType
```
Comment (blockchaindevsh) [inline op-e2e/faultproofs/util.go]: No need to introduce unnecessary diff here.

File: op-e2e/faultproofs/util.go [hunk 2]
```diff
@@ -56,6 +59,18 @@ func WithAllocType(allocType config.AllocType) faultDisputeConfigOpts {
 	}
 }
 
+func WithDAC(dacUrl string) faultDisputeConfigOpts {
+	return func(cfg *e2esys.SystemConfig) {
+		if c, ok := cfg.Nodes["sequencer"]; ok {
+			c.DACConfig = &node.DACConfig{URLS: []string{dacUrl}}
+			c.Driver = driver.Config{SequencerEnabled: true}
```
Comment (blockchaindevsh) [inline op-e2e/faultproofs/util.go]: It may override other configurations within `driver.Config`.

File: op-e2e/faultproofs/util.go [hunk 3]
```diff
@@ -56,6 +59,18 @@ func WithAllocType(allocType config.AllocType) faultDisputeConfigOpts {
...
+			c.Rollup.L2BlobConfig = &rollup.L2BlobConfig{
```
Comment (blockchaindevsh) [inline op-e2e/faultproofs/util.go]: It's not good abstraction to put it here, you can follow how `rollupAltDAConfig` works [here](https://github.com/ethstorage/optimism/blob/32380c23aa610429e82cf0e9f323e79e364687eb/op-e2e/system/e2esys/setup.go#L613).

File: op-e2e/system/e2esys/setup.go [hunk 4]
```diff
@@ -813,6 +813,9 @@ func (cfg SystemConfig) Start(t *testing.T, startOpts ...StartOption) (*System,
 		c.Rollup.LogDescription(cfg.Loggers[name], chaincfg.L2ChainIDToNetworkDisplayName)
 		l := cfg.Loggers[name]
 
+		if name == "sequencer" {
```
Comment (blockchaindevsh) [inline op-e2e/system/e2esys/setup.go]: This won't be necessary if done the same way as `rollupAltDAConfig`.

File: op-e2e/blobs/blob_test.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	for i := range blobs {
+		b := GetRandBlob(t, int64(i))
+		blobs[i] = &b
+	}
+
+	tx, err := sendTransactionWithBlobs(t, ctx, l2Client, sys.TestAccount(0), toAddress, blobs)
+	require.NoError(t, err)
+	_, err = wait.ForReceiptOK(ctx, l2Client, tx.Hash())
+
+	// TODO: wait for DAC server bug fix check in to refactor here
```
Comment (syntrust) [inline op-e2e/blobs/blob_test.go]: Bug to prevent blobs from being downloaded in batch?

File: op-e2e/blobs/blob_test.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	toAddress = testutils.RandomAddress(mrand.New(mrand.NewSource(dacPort)))
+	blobs     = make([]*eth.Blob, 3)
+)
+
+func TestFunctionSuccess(t *testing.T) {
+	op_e2e.InitParallel(t)
+	dacServer := StartDACServer(t)
+	defer dacServer.Stop(ctx)
+
+	sys, _ := faultproofs.StartFaultDisputeSystem(t, faultproofs.WithDAC(dacUrl))
```
Comment (syntrust) [inline op-e2e/blobs/blob_test.go]: Why do we need a dispute system from `faultproofs` here? Is there any relationship between L2 blob and fault proof in this case?

File: op-node/rollup/types.go [hunk 7]
```diff
@@ -685,6 +685,7 @@ func (c *Config) LogDescription(log log.Logger, l2Chains map[string]string) {
 		"holocene_time", fmtForkTimeOrUnset(c.HoloceneTime),
 		"interop_time", fmtForkTimeOrUnset(c.InteropTime),
 		"alt_da", c.AltDAConfig != nil,
+		"L2BlobConfig", c.L2BlobConfig != nil,
```
Comment (syntrust) [inline op-node/rollup/types.go]: `L2BlobConfig` better follow the "snake_case" style as others?

File: op-e2e/blobs/blob_test.go (hunk 5 above)
Comment (ping-ke) [reply op-e2e/blobs/blob_test.go]: yes

File: op-e2e/blobs/blob_test.go (hunk 6 above)
Comment (ping-ke) [reply op-e2e/blobs/blob_test.go]: no, just reuse this set up func just sgt e2e

File: op-e2e/system/e2esys/setup.go [hunk 8]
```diff
@@ -620,6 +620,13 @@ func (cfg SystemConfig) Start(t *testing.T, startOpts ...StartOption) (*System,
 		}
 	}
 
+	var l2BlobConfig *rollup.L2BlobConfig
+	if c, ok := cfg.Nodes["sequencer"]; ok && c.DACConfig != nil {
+		l2BlobConfig = &rollup.L2BlobConfig{
```
Comment (blockchaindevsh) [inline op-e2e/system/e2esys/setup.go:625]: `rollup.L2BlobConfig` should be configured by `DeployConfig.L2GenesisBlobTimeOffset`, not by `DACConfig`. In theory, user can set `rollup.L2BlobConfig` with or without `DACConfig`.

**Review by syntrust — APPROVED** [review]

File: op-e2e/blobs/blob_test.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+		if len(blob) != eth.BlobSize {
+			t.Error("Invalid downloaded blob len", "blob hash", blobHash, "blob len", len(blob))
+		}
+		if bytes.Compare(blob, blobs[i][:]) != 0 {
+			t.Error("blob content", blob[:32], blobs[i][:32])
+		}
+	}
+}
+
+func StartSystemWithDAC(t *testing.T) (*e2esys.System, *ethclient.Client) {
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Don't use capitalized first character unless you want to export it. The same for other functions.

File: op-e2e/blobs/blob_test.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
+package blobs
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Please move it to a directory named l2blobs to unify the naming conventions.

File: op-e2e/blobs/blob_test.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+const (
+	dacPort = 37777
+)
+
+var (
+	ctx       = context.Background()
+	dacUrl    = fmt.Sprintf("http://127.0.0.1:%d", dacPort)
+	toAddress = testutils.RandomAddress(mrand.New(mrand.NewSource(dacPort)))
+	blobs     = make([]*eth.Blob, 3)
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Why not just define it inside `TestFunctionSuccess` since it's only referenced in that function? The same for `toAddress`.

File: op-e2e/blobs/blob_test.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	"github.com/holiman/uint256"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	dacPort = 37777
+)
+
+var (
+	ctx       = context.Background()
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: It's recommended to use a context that will be canceled within seconds instead of never canceled.

File: op-e2e/blobs/blob_test.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	delete(cfg.Nodes, "verifier")
+	if c, ok := cfg.Nodes["sequencer"]; ok {
+		c.SafeDBPath = t.TempDir()
+		c.DACConfig = &node.DACConfig{URLS: []string{dacUrl}}
+		c.Driver.SequencerEnabled = true
+	}
+	cfg.DeployConfig.L2GenesisBlobTimeOffset = new(hexutil.Uint64)
+	cfg.DeployConfig.SequencerWindowSize = 30
+	cfg.DeployConfig.FinalizationPeriodSeconds = 2
+	cfg.SupportL1TimeTravel = true
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Please add a comment about why the changes to `SequencerWindowSize`/`FinalizationPeriodSeconds`/`SupportL1TimeTravel` are necessary.

File: op-e2e/blobs/blob_test.go [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+		if bytes.Compare(blob, blobs[i][:]) != 0 {
+			t.Error("blob content", blob[:32], blobs[i][:32])
+		}
+	}
+}
+
+func StartSystemWithDAC(t *testing.T) (*e2esys.System, *ethclient.Client) {
+	cfg := e2esys.DefaultSystemConfig(t)
+	delete(cfg.Nodes, "verifier")
+	if c, ok := cfg.Nodes["sequencer"]; ok {
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: I think here you should assert "sequencer" exists in the map, it's not optional.

File: op-e2e/blobs/blob_test.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+		new(big.Int).Mul(head.BaseFee, big.NewInt(2)),
+	)
+	return tip, gasFeeCap, blobFee, nil
+}
+
+func GetRandBlob(t *testing.T, seed int64) eth.Blob {
+	r := mrand.New(mrand.NewSource(seed))
+	bigData := eth.Data(make([]byte, eth.MaxBlobDataSize))
+	for i := range bigData {
+		bigData[i] = byte(r.Intn(256))
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: You can use `r.Read()` to fill `bigData` in 1 go.

File: op-e2e/blobs/blob_test.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+	_, err = wait.ForReceiptOK(ctx, l2Client, tx.Hash())
+
+	dblobs, err := downloadBlobs(dacUrl, tx.BlobHashes())
+	require.NoError(t, err)
+	require.True(t, len(dblobs) == len(tx.BlobHashes()), "blobs downloaded is not equal to blob hashes")
+
+	for i, blobHash := range dblobs {
+		blob := dblobs[i]
+		if len(blob) != eth.BlobSize {
+			t.Error("Invalid downloaded blob len", "blob hash", blobHash, "blob len", len(blob))
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Please always use `require.xxx` for consistency.

File: op-e2e/blobs/blob_test.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		if bytes.Compare(blob, blobs[i][:]) != 0 {
+			t.Error("blob content", blob[:32], blobs[i][:32])
```
Comment (blockchaindevsh) [inline op-e2e/blobs/blob_test.go]: Please always use `require.xxx` for consistency.

File: op-e2e/l2blobs/l2blob_test.go [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+			common.Bytes2Hex(blob[:32]), common.Bytes2Hex(blobs[i][:32])))
+	}
+}
+
+func startSystemWithDAC(t *testing.T) (*e2esys.System, *ethclient.Client) {
+	cfg := e2esys.DefaultSystemConfig(t)
+	delete(cfg.Nodes, "verifier")
+	c, ok := cfg.Nodes["sequencer"]
+	require.True(t, ok, "sequencer is required")
+	if ok {
```
Comment (blockchaindevsh) [inline op-e2e/l2blobs/l2blob_test.go]: No need for `if` since ok is guaranteed to be true here.

File: op-e2e/l2blobs/l2blob_test.go [hunk 19]
```diff
@@ -1,4 +1,4 @@
-package blobs
+package l2blobs
```
Comment (zhiqiangxu) [inline op-e2e/l2blobs/l2blob_test.go:1]: Let's move it to `l2blob` so that the naming convention is consistent with `l2blob_test.go`.

File: op-e2e/l2blobs/l2blob_test.go [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
```
Comment (blockchaindevsh) [inline op-e2e/l2blobs/l2blob_test.go]: Let's move it to `l2blob` so that the naming convention is consistent with `l2blob_test.go`.

**Review by blockchaindevsh — APPROVED** [review]
LGTM assuming the last two comments are resolved.
